sunscreen/tests/linked.rs

```diff
@@ -162,6 +163,150 @@ mod linked_tests {
+            let mut proof_builder = LogProofBuilder::new(&rt);
+            let (_ct, x_msg) = proof_builder.encrypt_and_link(&x, &public_key).unwrap();
+            proof_builder
+                .zkp_program(is_eq_zkp)
+                .unwrap()
+                .linked_input(&x_msg)
+                .private_input(BulletproofsField::from(x_n))
+                .private_input(BulletproofsField::from(x_d));
+
+            let lp = proof_builder
+                .build_linkedproof()
+                .unwrap_or_else(|e| panic!("Failed to prove encoding of {x:?}; {e}"));
+            lp.verify::<ZkpProgramInput>(is_eq_zkp, vec![], vec![])
+                .unwrap_or_else(|e| panic!("Failed to verify encoding of {x:?}; {e}"));
+        }
+    }
+
+    #[zkp_program]
+    fn compare_signed<F: FieldSpec>(#[linked] x: BfvSigned<F>, #[linked] y: BfvSigned<F>) {
+        x.into_field_elem()
+            .constrain_le_bounded(y.into_field_elem(), 64)
+    }
+
+    #[test]
+    fn can_compare_signed() {
+        let app = Compiler::new()
+            .fhe_program(doggie)
+            .with_params(&SMALL_PARAMS)
+            .zkp_backend::<BulletproofsBackend>()
+            .zkp_program(compare_signed)
+            .compile()
+            .unwrap();
+        let rt = FheZkpRuntime::new(app.params(), &BulletproofsBackend::new()).unwrap();
+        let compare_signed_zkp = app.get_zkp_program(compare_signed).unwrap();
+
+        let (public_key, _secret_key) = rt.generate_keys().unwrap();
+
+        // To slow to run in a loop :/ if we eventually expose small params for testing, do more
+        // cases
+        for _ in 0..1 {
+            let x = rand::random::<i32>() as i64;
+            let y = rand::random::<i32>() as i64;
+            let (x, y) = (Signed::from(x.min(y)), Signed::from(x.max(y)));
+
+            let mut proof_builder = LogProofBuilder::new(&rt);
+            let (_ct, x_msg) = proof_builder.encrypt_and_link(&x, &public_key).unwrap();
+            let (_ct, y_msg) = proof_builder.encrypt_and_link(&y, &public_key).unwrap();
+            proof_builder
+                .zkp_program(compare_signed_zkp)
+                .unwrap()
+                .linked_input(&x_msg)
+                .linked_input(&y_msg);
+
+            let lp = proof_builder
+                .build_linkedproof()
+                .unwrap_or_else(|e| panic!("Failed to prove {x:?} <= {y:?}, {e}"));
+            lp.verify::<ZkpProgramInput>(compare_signed_zkp, vec![], vec![])
+                .unwrap_or_else(|e| panic!("Failed to verify {x:?} <= {y:?}; {e}"));
+        }
+    }
+
+    #[zkp_program]
+    fn compare_rational<F: FieldSpec>(#[linked] x: BfvRational<F>, #[linked] y: BfvRational<F>) {
+        let (x_n, x_d) = x.into_field_elem();
+        let (y_n, y_d) = y.into_field_elem();
+        let x = x_n * y_d;
+        let y = y_n * x_d;
+        x.constrain_le_bounded(y, 64)
+    }
+
+    #[test]
+    #[ignore = "bug in linked input multplication"]
```

Comment:
This issue was discovered in these tests, but it is preexisting and independent of the work in this PR. Will be tackled next week.

Reply:
Ah I see! I bet the multiplication is throwing off the assumed generator order by adding a new input prior to where the link component thinks the multipliers will be.